returnn/util/file_cache.py

```diff
@@ -203,45 +200,64 @@ def cleanup(self, *, need_at_least_free_space_size: int = 0):
         for mtime, neg_size, fn in all_files:
             size = -neg_size
             delete_reason = None
-            if cur_time - mtime > self._cleanup_files_always_older_than_days * 60 * 60 * 24:
-                delete_reason = f"File is {(cur_time - mtime) / 60 / 60 / 24:.1f} days old"
-            else:
-                reached_more_recent_files = True
-            if not delete_reason and need_at_least_free_space_size > cur_expected_free:
-                # Still must delete some files.
-                if cur_time - mtime > cur_used_time_threshold:
-                    delete_reason = f"Still need more space, file is {(cur_time - mtime) / 60 / 60:.1f} hours old"
+
+            lock_dir, lock_file = self._get_lock_filename(fn)
+            with LockFile(directory=lock_dir, name=lock_file, lock_timeout=self._lock_timeout):
+                # Re-check mtime with lock, could have been updated by another
+                # process in the meantime.
+                #
+                # We do not update the `mtime` variable here, because the code
+                # assumes that the list of files is sorted by mtime to abort
+                # early when enough space has been made.
+                #
+                # Instead, we treat the case where the mtime was updated during
+                # cleanup as an outlier and continue as if no other mtimes had
+                # changed.
```

Comment:
```suggestion
                # Re-check mtime with lock, could have been updated by another
                # process in the meantime.
                # We do not update the `mtime` variable here, because the code
                # assumes that the list of files is sorted by mtime to abort
                # early when enough space has been made.
                # Instead, we treat the case where the mtime was updated during
                # cleanup as an outlier and continue as if no other mtimes had
                # changed.
                # See for discussion:
                #   - https://github.com/rwth-i6/returnn/issues/1675
                #   - https://github.com/rwth-i6/returnn/pull/1709
```